00_homework_kawapa/Z9/2-morph-line-detection/main.cpp

```diff
@@ -0,0 +1,98 @@
+#include <opencv2/core.hpp>
+#include <opencv2/imgproc.hpp>
+#include <opencv2/highgui.hpp>
+#include <array>
+#include <future>
+#include <functional>
+#include <iostream>
+#include <vector>
+
+using namespace cv;
+
+constexpr size_t n = 4; // number of kernels
+std::array<cv::Scalar, n> colours {(255,0,0), (0, 255, 0), (0, 0, 255), (255, 255, 0)};
+
+void extractLines(Mat & inputOutput, const Mat & kernel);
+void prepare(const Mat &, Mat &, Mat &);
+void show_wait_destroy(const char* winname, cv::Mat img);
+
+int main(int argc, char** argv)
+{
+    Mat gray;
+    Mat bw;
+    
+    CommandLineParser parser(argc, argv, "{@input | notes.png | input image}");
+    Mat src = imread(samples::findFile(parser.get<String>("@input")), IMREAD_COLOR);
+
+    if (src.empty())
+    {
+        std::cout << "Could not open or find the image!\n" << std::endl;
+        std::cout << "Usage: " << argv[0] << " <Input image>" << std::endl;
+        return -1;
+    }
+
+    prepare(src, gray, bw);
+    
+    std::array<const char*, n> directions {"horizontal", "vertical", "diagonalLeft", "diagonalRight"};
+    std::vector<std::future<void>> futures(n);
+
+    std::vector<cv::Mat> results(n);
+    std::for_each(results.begin(), results.end(), [=](Mat & obj){ obj = bw.clone(); });
+    
+    std::vector<cv::Mat> kernels;
+    kernels.reserve(n);
```

Comment:
to warto przenieść dopiero do linii 54. Wtedy można od razu utworzyć wektor kernels w właściwymi wartościami na starcie, bez emplace_back. To wszystko może być też przerzucone do funkcji, która zwróci taki wektor.